agent/lib/kontena/network_adapters/dns_manager.rb

```diff
@@ -2,177 +2,99 @@
-        self.start_container(container)
+        self.register_container_dns(container)
       end
     end
 
+    def started?
+      !!@started
+    end
+
+    # @param topic [String]
+    # @param event [Docker::Event]
     def on_dns_add(topic, event)
       wait_weave_running?
       add_dns(event[:id], event[:ip], event[:name])
     end
 
-    # @param [String] topic
-    # @param [Docker::Event] event
+    # @param topic [String]
+    # @param event [Docker::Event]
     def on_container_event(topic, event)
-      # cannot run start_container before start has populated @migrate_containers
       return unless started?
 
-      if event.status == 'start'
+      if event.status == 'start'.freeze
         container = Docker::Container.get(event.id) rescue nil
         if container
-          self.start_container(container)
+          register_container_dns(container) if container.infra_container?
         else
           warn "skip start event for missing container=#{event.id}"
         end
-      elsif event.status == 'restart'
-        if router_image?(event.from)
-          wait_weave_running?
-
-          self.start
-        end
-      elsif event.status == 'destroy'
-        self.on_container_destroy(event)
       end
-    end
-
-    def started?
-      @started
-    end
-
-    # Ensure weave network for container
-    #
-    # @param [Docker::Container] container
-    def start_container(container)
-      overlay_cidr = container.overlay_cidr
-
-      if overlay_cidr
-        wait_weave_running?
-
-        register_container_dns(container)
-        attach_overlay(container)
-      else
-        debug "skip start for container=#{container.name} without overlay_cidr"
-      end
-    rescue Docker::Error::NotFoundError
-      debug "skip start for missing container=#{container.id}"
-
     rescue => exc
-      error "failed to start container: #{exc.class.name}: #{exc.message}"
-      error exc.backtrace.join("\n")
+      error exc
     end
 
-    # @param [Docker::Event] event
-    def on_container_destroy(event)
-      container_id = event.id
-      overlay_network = event.Actor.attributes['io.kontena.container.overlay_network']
-      overlay_cidr = event.Actor.attributes['io.kontena.container.overlay_cidr']
-
-      if overlay_network && overlay_cidr
-        wait_network_ready?
-
-        network_adapter.remove_container(container_id, overlay_network, overlay_cidr)
-      end
-    rescue => exc
-      error "failed to remove container: #{exc.class.name}: #{exc.message}"
-      error exc.backtrace.join("\n")
-    end
-
-    OVERLAY_SUFFIX = '16'
-
-    # @param [String] container_id
-    # @param [String] overlay_cidr
-    def attach_overlay(container)
-      if container.overlay_network.nil?
-        # overlay network migration for 0.16 compat
-        # override overlay network /19 -> /16 suffix for existing containers that may need to be migrated
-        overlay_cidr = "#{container.overlay_ip}/#{OVERLAY_SUFFIX}"
```

Comment:
Time to drop the 0.16 compatibility?

Reply:
We need to migrate containers anyway so no need to keep legacy.

Reply:
:+1: perfectly okay with this, being able to migrate containers across upgrades would spare us a lot of backwards-compatibility pain. But it will need something to coordinate the container upgrades to not be too disruptive...